mesh/ncmesh.hpp

```diff
@@ -32,17 +32,35 @@ namespace mfem
 /** Represents the index of an element to refine, plus a refinement type.
     The refinement type is needed for anisotropic refinement of quads and hexes.
     Bits 0,1 and 2 of 'ref_type' specify whether the element should be split
-    in the X, Y and Z directions, respectively (Z is ignored for quads). */
+    in the X, Y and Z directions, respectively (Z is ignored for quads). The
+    refinement spacing or scale in each direction is a number in (0,1), with the
+    default 0.5 meaning bisection. This linear scale parameter defines the
+    position of the refinement between the beginning (0) and end (1) of the
+    reference element in each direction. */
 struct Refinement
 {
    enum : char { X = 1, Y = 2, Z = 4, XY = 3, XZ = 5, YZ = 6, XYZ = 7 };
    int index; ///< Mesh element number
    char ref_type; ///< refinement XYZ bit mask (7 = full isotropic)
+   real_t scale_x, scale_y, scale_z; ///< Scaling in each direction
 
    Refinement() = default;
 
-   Refinement(int index, int type = Refinement::XYZ)
-      : index(index), ref_type(type) {}
+   Refinement(int index, int type = Refinement::XYZ, real_t s_x = 0.5,
+              real_t s_y = 0.5, real_t s_z = 0.5)
```

Comment:
If the user wants to refine in Z only, will they have to specify dummy values for `s_x` and `s_y`?

Reply:
Yes, unfortunately the user will have to specify dummy values for directions in which there is no refinement, and I agree that this is awkward. Your suggestion would be convenient for the 3 cases of refinement in one direction, but there are also the 3 refinement types `XY`, `XZ`, and `YZ`, which also require a dummy value and 2 real values. For those types, `std::pair` would not work. We can make functions like `SetX`, `SetXY`, etc., but I think this constructor is still more convenient, despite the dummy values. Another idea is to make a `struct RefinementType` to replace the `enum`. We could have all types derive from that base struct, e.g. `RefinementX`, `RefinementXY`, etc. That way, this constructor for `Refinement` would take a `RefinementType` and then get the values `s_x`, `s_y`, `s_z` from the derived struct, which would set the dummy values automatically.

Reply:
> Your suggestion would be convenient for the 3 cases of refinement in one direction, but there are also the 3 refinement types XY, XZ, and YZ, which also require a dummy value and 2 real values. For those types, std::pair would not work.

FWIW, I thought the XY, XZ, and YZ refinements could be constructed with `Refinement(index, {{'X',0.5},{'Y",0.5}})`, `Refinement(index, {{'X',0.5},{'Z",0.5}})`, and `Refinement(index, {{'Y',0.5},{'Z",0.5}})`, respectively.  Or perhaps there's an issue I don't yet see.  That's not to say introducing a hierarchy of structs isn't better in some way... I think either my suggestion or the hierarchy of structs will provide a cleaner API to the user.

Reply:
Right, we could use a list of refinements in each direction, instead of the `enum`, as the type can be inferred from the directions included in the list. I agree that (i) the list of refinements in each direction and (ii) the derived structs are both good options, better than the current version. Such API questions often get lots of input, so let's see if anyone else has opinions.

Reply:
I changed the interface in a926841cf9eeb307f0cdaa48adc51b8f2f096926 similarly to the code suggested by @hughcars. One difference is that I had to use `real_t s[3]` to keep `Refinement` as a trivial type for the `Array` template type. The default refinement, using an empty list, requires `, {}`, which is a little inconvenient but not as bad as alternatives.

Reply:
Commit https://github.com/mfem/mfem/pull/4371#:~:text=the%20interface%20in-,a926841,-similarly%20to%20the breaks backwards compatibility in 2 ways. One is the `, {}` for an empty list to get the default type, and the other is the enum name `Type`. We may be able to avoid the `, {}` by rearranging arguments in constructors, or by using a set function instead of a constructor with a list, but the enum name `Type` seems necessary in order to use the pair of type and scale. For example, `Refinement::Type::X` is necessary now, instead of just `Refinement::X`. I'm not sure we should make this breaking change here. Maybe it is something to keep in mind for MFEM 5.0. As the editor, what do you think @tzanio?

Reply:
The need for `Type::` comes from `enum` becoming `enum class`. Generally `enum class` is preferred as it prevents accidental aliasing of the underlying type into an `enum` by accident. That can be removed though if need be, but I would suggest it's good practice to use it. Adding another constructor that takes only the integer value and dispatches to that constructor with `, {}` would fix that issue: `Refinement(int i) : Refinement(i, {}) {}` in the class body should suffice.

Reply:
I was thinking that `Type` would be necessary in the `std:pair`, but I was able to replace it with `char` in e5de10c9d91b1b2f6d8a8915bbf117baa5f96d7b which restores backwards compatibility.

Reply:
Even though I wasn't very active in this thread, I suppose it's on me as the starter of the thread to resolve it... please feel free to re-open if neeeded